src/bit-systems/link-hover-menu.ts

```diff
@@ -1,47 +1,150 @@
+      const waypoint = url.hash && url.hash.substring(1);
+      // move to new room without page load or entry flow
+      const hubId = hubIdFromUrl(url);
+      changeHub(hubId, true, waypoint);
+      break;
+    case LinkType.ROOM_URL:
+      await exitImmersive();
+      location.href = src;
+      break;
+  }
+}
+
+function moveToTarget(world: HubsWorld, menu: EntityID) {
+  const linkEid = LinkHoverMenu.targetObjectRef[menu];
+  const targetObject = world.eid2obj.get(linkEid)!;
+  targetObject.updateMatrices();
+  const menuObject = world.eid2obj.get(menu)!;
+  setMatrixWorld(menuObject, targetObject.matrixWorld);
+}
+
+function updateButtonText(world: HubsWorld, menu: EntityID, button: EntityID) {
+  const text = findChildWithComponent(world, TextTag, button)!;
+  const textObj = world.eid2obj.get(text)! as TroikaText;
+  const linkEid = LinkHoverMenu.targetObjectRef[menu];
+  const linkType = Link.type[linkEid];
+  let label = "open link";
+  switch (linkType) {
+    case LinkType.AVATAR:
+      label = "use avatar";
+      break;
+    case LinkType.SCENE:
+      label = "use scene";
+      break;
+    case LinkType.WAYPOINT:
+      label = "go to";
+      break;
+    case LinkType.ROOM:
+    case LinkType.ROOM_URL:
+      label = "visit room";
+      break;
+  }
+  textObj.text = label;
+}
+
+function flushToObject3Ds(world: HubsWorld, menu: EntityID, frozen: boolean) {
+  const target = LinkHoverMenu.targetObjectRef[menu];
+  const visible = !!(target && !frozen);
 
-  // Check if the cursor it hovered on Link object or Link menu button object.
-  const hovered = hoveredQuery(world).length > 0 || hoveredMenuItemQuery(world).length > 0;
-
-  const linkEid = LinkHoverMenu.targetObjectRef[menuEid];
-  if (hovered && entityExists(world, linkEid)) {
-    // Hovered then make the menu visible and handle clicks if needed.
-    menuObject.visible = true;
-    if (clickedMenuItemQuery(world).length > 0) {
-      // TODO: Change the action and label depending on 
-      //       target (eg: switching avatars for avatar url)
-      window.open(APP.getString(Link.url[linkEid])!);
+  const obj = world.eid2obj.get(menu)!;
+  obj.visible = visible;
+
+  const linkButtonRef = LinkHoverMenu.linkButtonRef[menu];
+  const buttonObj = world.eid2obj.get(linkButtonRef)!;
+  // Parent visibility doesn't block raycasting, so we must set each button to be invisible
+  // TODO: Ensure that children of invisible entities aren't raycastable
+  if (visible) {
+    if (visible !== buttonObj.visible) {
+      updateButtonText(world, menu, linkButtonRef);
+      buttonObj.visible = true;
     }
   } else {
-    // Not hovered or target object has already been deleted
-    // then make the menu invisible and forget the Link object.
-    menuObject.visible = false;
-    if (menuObject.parent !== null) {
-      menuObject.parent.remove(menuObject);
-    }
-    LinkHoverMenu.targetObjectRef[menuEid] = NULL_EID;
+    buttonObj.visible = false;
   }
```

Comment:
If I'm right, unexpectedly text won't be updated in this scenario, will be?

1. A Link entity is hovered and button is visible
2. At the next frame, another different typed Link entity is hovered
3. Text needs to be updated because of different link type
4. But text is not updated because button is visible in both this frame and the previous frame

Reply:
Good catch, that's right, we need to check for target updates too.